Accumulator & Evaluation strategy

wolfv opened the pull request:
I am planning to add tests + immediate evaluation for reducers today.

wolfv commented:
fixes #474

JohanMabille commented on include/xtensor/xaccumulator.hpp:
```diff
@@ -0,0 +1,133 @@
+        xarray<typename std::decay_t<E>::value_type> accumulator_impl(F&& f, E&& e, std::size_t axis, evaluation_strategy::immediate)
+        {
+            using T = typename std::decay_t<E>::value_type;
+            std::size_t sz = compute_size(e.shape());
```
As pointed out by codacy, `sz` is never used

JohanMabille commented on include/xtensor/xaccumulator.hpp:
```diff
@@ -0,0 +1,133 @@
+            for (; i < axis; ++i)
+            {
+                outer_loop_dim *= result.shape()[i];
+            }
```
```cpp
std::size_t outer_loop_dim = std::accumulate(result.shape().cbegin(), result.shape().cbegin() + std::size_t(axis), std::size_t(1), std::multiplies());
```

could replace the loop

JohanMabille commented on include/xtensor/xaccumulator.hpp:
```diff
@@ -0,0 +1,133 @@
+            for (; i < result.dimension(); ++i)
+            {
+                inner_loop_dim *= result.shape()[i];
+            }
```
As for the previous loop, `std::accumulate` could be used here.

JohanMabille commented on include/xtensor/xexception.hpp:
```diff
@@ -92,6 +92,26 @@ namespace xt
+     * not_implemented_exception *
+     *****************************/
+
+    class not_implemented_exception : public std::logic_error
```
I think this can be removed since you use static_assert for non implemented stuff.

JohanMabille commented:
After a rebase and removing the unused `not_implemented_exception`, I think this can be merged.

wolfv commented:
@JohanMabille I have rebased and squashed a couple of commits, but tests for the greedy reducer are still missing (I noticed).

Is there an easy way to template the xreducer tests to run with greedy and lazy? :)

JohanMabille commented:
@wolfv you can use the TYPED_TEST macro of google test (it is already extensively used for operation and semantic tests).

That would give something like
```cpp
using testing_types = ::testing::Types<evaluation_strategy::immediate, evaluation_strategy::lazy>;
TYPED_TEST_CASE(xreducer, testing_types);

TYPED_TEST(xreducer, name_of_your_test)
{
    // initialization of everything
    auto res = xt::reduce(f, e, axes, TypeParam());
    // ...
}
```

wolfv commented:
@JohanMabille if you could have another look ... unfortunately this PR has grown a bit in scope ... but should be good now! 

Benchmarks look quite nice (the immediate_reducer is the one added in this PR, and the strided reducer is a not very functional simple pure-C++ reducer)

```
---------------------------------------------------------------------------------------
Benchmark                                                Time           CPU Iterations
---------------------------------------------------------------------------------------
benchmark_reducer/10x100000/axis 0                 1492081 ns    1481461 ns        400
benchmark_reducer/10x100000/axis 1                 1253155 ns    1245567 ns        434
benchmark_reducer/100000x10/axis 1                 4182459 ns    4143760 ns        153
benchmark_reducer/100000x10/axis 0                 1545567 ns    1537210 ns        362
benchmark_reducer/100000x10/axis both              1219680 ns    1216229 ns        429
benchmark_immediate_reducer/10x100000/axis 0        809790 ns     805997 ns        862
benchmark_immediate_reducer/10x100000/axis 1       1173746 ns    1170508 ns        453
benchmark_immediate_reducer/100000x10/axis 1        814415 ns     811736 ns        863
benchmark_immediate_reducer/100000x10/axis 0        834972 ns     832468 ns        820
benchmark_immediate_reducer/100000x10/axis both    1153821 ns    1151137 ns        469
benchmark_strided_reducer/10x100000/axis 0          721150 ns     718172 ns        917
benchmark_strided_reducer/10x100000/axis 1         1152899 ns    1149744 ns        615
benchmark_strided_reducer/100000x10/axis 1         5344045 ns    5299108 ns         97
benchmark_strided_reducer/100000x10/axis 0          796759 ns     793212 ns        905
```

JohanMabille commented on include/xtensor/xreducer.hpp:
```diff
@@ -37,20 +37,231 @@ namespace xt
+            }
+            return std::make_pair(i == -1,
+                                  std::inner_product(temp_idx.begin(), temp_idx.end(),
+                                                     iter_strides.begin(), ptrdiff_t(0)));
```
Ok the vector of pair may be annoying here; you need an additional `xindex` before the lambda definition to store an intermediate result; since we removed a dynamic allocation when we merged `iter_shape` and `iter_strides`, this won't be a performance issue compared to the initial implementation. Let's call it `tmp_prod`, you you would have to replace `std::inner_product` with:
```cpp
std::transform(temp_idx.begin(), temp.idx.end(), iter_shape_strides.begin(), tmp_prod.begin(), auto[](auto&& t1, auto&& t2) { return t1 * t2.second; });
return std::make_pair(i == -1,
std::accumulate(tmp_prod.begin(), tmp_prod.end(), ptrdiff_t(0)));
```
This less expressive than `inner_product`, but it is inside a lambda whose name is expressive enough, and it simplifies the code before so I think it is worth doing it.

wolfv replied:
I think there is a more straightforward way to initialize the iter_shape & iter_strides, without having to remove any elements by just filling in the right ones. But I kind of had to go back to the drawing board with this function and derive it logically before making such optimizations.

I'll try to add that later in this PR.

JohanMabille commented on include/xtensor/xreducer.hpp:
```diff
@@ -37,20 +37,231 @@ namespace xt
+        }
+
+        xindex temp_idx(iter_shape.size());
+        auto next_idx = [&iter_shape, &iter_strides, &temp_idx]()
```
why capturing `temp_idx` instead of passing it as const reference ?

wolfv replied:
temp_idx holds the state of the iteration and is modified in the loop. you mean instead as a function argument?

JohanMabille replied:
yep sorry I mean as an argument of the next_id function. And indeed it should be a simple reference, not a const reference. It's just that I find it clearer to have (but that's totally subjective)
```cpp
auto res = next_idx(temp_idx);
```

wolfv replied:
the thing is that temp_idx is basically internal to the functor/lambda here, and isn't used outside. that's why i decided to do it like this

JohanMabille commented on include/xtensor/xreducer.hpp:
```diff
@@ -37,20 +37,231 @@ namespace xt
+                else
+                {
+                    *out = tmp;
+                }
```
maybe replace with `*out = merge ? merge_fct(*out, tmp) : tmp;`
The code is long enough so that using a ternary conditional operator is justified.

wolfv replied:
:+1:

JohanMabille commented on include/xtensor/xreducer.hpp:
```diff
@@ -37,20 +37,231 @@ namespace xt
+                    {
+                        *(out + i) = merge_fct(*(out + i), *(begin + i));
+                    }
+                }
```
Lines from 220 to 233 can be simplified with
```cpp
std::transform(out, out + inner_loop_size, begin, out, [merge, &init_fct, &merge_fct](auto&& v1, auto&& v2) { return merge ? merge_fct(v1, v2) : init_fct(v2); });
```

JohanMabille commented on include/xtensor/xreducer.hpp:
```diff
@@ -37,20 +37,231 @@ namespace xt
+                    for (std::size_t j = 0; j < inner_loop_size; ++j)
+                    {
+                        *(out + j) = acc_fct(*(out + j), *(begin + j));
+                    }
```
This can be simplified with
```cpp
std::transform(out, out + inner_loop_size, begin, out, acc_fct);
```

wolfv replied:
too bad the unary std::transform doesn't exist yet! we could totally use it to parallelize everything and add a `execution_policy` tag! :) C++17, here we come!

JohanMabille commented on include/xtensor/xsort.hpp:
```diff
@@ -0,0 +1,411 @@
+
+        template <class T>
+        class strided_iterator :
+            public std::iterator<std::random_access_iterator_tag, T, ptrdiff_t, const T*, T&>
```
I don't understand the need for this. Why not using steppers ? (they basically do the same and work with any kind of expession, not only the strided ones).

I think this deserves a dedicated PR.

wolfv replied:
Is there a way to give the stepper a RandomIterator interface (e.g. `++` and `+(int)`)? Instead of `step(steps, dim)`?
I'd love to remove all this code!

JohanMabille replied:
Let's discuss this in the PR dedicated to `xsort.hpp`.

JohanMabille commented:
As said in the review, I think the `xsort.hpp` file deserves a dedicated PR. I'm waiting for it before starting a review, so we can go ahead on the accumulator feature.

wolfv commented on include/xtensor/xaccumulator.hpp:
```diff
@@ -0,0 +1,136 @@
+            {
+                result.data()[i + 1] = f(result.data()[i], e.data()[i + 1]);
+            }
+            return std::move(result);
```
Done!

wolfv commented on include/xtensor/xreducer.hpp:
```diff
@@ -37,20 +37,231 @@ namespace xt
+        auto next_idx = [&iter_shape, &iter_strides, &temp_idx]()
+        {
+            int i = iter_shape.size() - 1;
+            for (; i >= 0; --i)
```
yeah! saving some performance!

wolfv replied:
or not, but much nicer ;)

JohanMabille commented on include/xtensor/xreducer.hpp:
```diff
@@ -128,36 +128,36 @@ namespace xt
+                iter_strides.erase(iter_strides.begin() + (ptrdiff_t) i);
                 ++i;
             }
         }
```
There's still a problem though; let's say you erase `iter_shape.begin()` (because `i = 0), when starting next iteration, `i = 2`, you never check `i = 1`. Besides, `iter_shape[0]` is now what was previously `iter_shape[1]`, so you never check it either.

wolfv replied:
oh, wow! thanks for noticing this. it's weird that the tests pass ... will definitely require a fix.

(I have an implementation of all this logic in Python and did tests against all axis combinations vs numpy -- we could also use that implementation to make some nice animations and teach people how to do fast reductions in the ND case ;)

wolfv commented:
closing in favor of #550 & #549
